server/plugin/utils.go

```diff
@@ -493,3 +494,9 @@ func (p *Plugin) CheckValidChannelForSubscription(channelID, userID string) (int
 
 	return 0, nil
 }
+
+func (p *Plugin) SanitizeURLPath(urlPath string) string {
+	parsedPath := strings.ReplaceAll(strings.ReplaceAll(urlPath, "..", ""), "%2e%2e", "")
```

Comment:
This covers the simple `..` and `%2e%2e` cases, but by "combinations of the two" I meant we also need to cover things like `.%2e`, `%2e.`, `.%2e%2e.`, `.%2e%2e%2e` and so forth.

Reply:
tried with regex to check for different combinations of `.` and `%2e` to eliminate path traversal patterns like `../`, `./`, `/../..`, etc
also if there is any handy doc can you please let me know what are the different harmful characters and their combination that we should prevent putting in a URL

Reply:
also Azure Devops allow putting dots in between other characters in their project name like `dummy..1..2..3..project` so, handled this also

Reply:
Since you're now matching `../` instead of just `..` like you previously did, the encoded form of the forward slash also becomes relevant, i.e `..%2f` and all other variations containing `%2f` need to be taken care of.

> also if there is any handy doc, can you please let me know?

There is no good documentation that you could directly apply because every scenario is a little different and these ones where the path traversal affects a server-side REST API call is a somewhat poorly recognized one in particular. You can find some general background info on [Wikipedia](https://en.wikipedia.org/wiki/Directory_traversal_attack) or the [OWASP Testing Guide](https://github.com/OWASP/wstg/blob/master/document/4-Web_Application_Security_Testing/05-Authorization_Testing/01-Testing_Directory_Traversal_File_Include.md), though, but don't take it at face value.

As you've noticed, the regexp/replace approach, especially applied to the completed URL instead of the inputs, gets very complicated very quickly and it's easy to make subtle mistakes. That's why my [original recommendation](https://github.com/mattermost/mattermost-plugin-azure-devops/pull/149#discussion_r1137684782) was to sanitize the inputs instead *before* incorporating them in the API path. It allows you to make better use of the stdlib `path.Clean` and `net/url.PathUnescape` functions and ditch the regexes.

Although I think @m1lt0n was the one to [suggest the current approach](https://github.com/mattermost/mattermost-plugin-azure-devops/pull/149#discussion_r1138527959), so maybe he has some other insights.